### DISCUSSION TITLE
dependency on pandas 1.0.0 -> please support older pandas as well

### DISCUSSION THREAD

**Discussion by loldenburg** [pitchmuc/adobe_analytics_api_2.0#62] [General]
Hi @pitchmuc 
when trying to update to aanalytics2 0.2.8,
I saw that it now depends on pandas>1.0.0.
This would require massive changes for me all over my code, tests etc...
I don't know if there were really changes added to aanalytics2 lately that require pandas 1.0.0. 
So if there is a way to make this not dependent on anything higher than pandas 0.25.3 (as it used to be), this would massively help me.
Thanks.

Lukas

**Comment by pitchmuc** [comment]
Hello Lukas,

There is no problem to reverse requirement to pandas 0.25.3. 
I was trying to clean up some legacy elements and upgrade the dependencies. 

I plan to do more "processing methods" in the future.

Example: a method to fetch the timeseries of an event and then forecast the expected value and compare it with reality. 
I would like to start building some functionalities like this with the "newest" version of pandas. 

This is not in place yet, so it can wait for a bit.
I will set back the requirements in the next release to 0.25.3.

**Comment by loldenburg** [comment]
Thanks a lot! Maybe I would not stretch the scope of this “wrapper” library
too far. Imho It should mainly be a wrapper and not be too heavy. For
processing this enhancements, awesome idea, but maybe that would be done in
a separate library which has this lib as a dependency

**Reply by pitchmuc** [reply]
Yes. I am thinking to create another library for this or just tutorials from that one wrapper. 
However, having pandas > 1.0.0 starts to get requirement for some libraries. 
The way that the dataframe structure the "None" values is a bit different (and other improvement on data manipulation). 

As pandas got its official release, I would encourage you to start testing for that one. 
It gets more and more recognized as the minimum library version.

**Reply by Ryuno-Ki** [reply]
Would environment markers be a way out?
https://www.python.org/dev/peps/pep-0508/#environment-markers

This way, you could load the old pandas version only for some older version of the Python interpreter.
Your code needs to be able to handle both variants, though.

**Reply by loldenburg** [reply]
@Ryuno-Ki thanks, I hadn't heard of these. The problem in my case at least is not different python interpreter versions. 
The problem is a whole repository (running on Python 3.8+) with 50 or more python scripts using pandas needs to be made compatible to pandas 1+.
But maybe I misunderstood the environment markers.

**Reply by Ryuno-Ki** [reply]
If you scroll down a bit you'll find a table of what criteria you can apply for environment markers.
I used it to support a Python 2 downstream project in a project I maintain (thus the example).

If you could figure out, whether there is some criteria that makes your „legacy” (as in: not-yet-Pandas-1+ ready) project from others, @pitchmuc could then declare Pandas <1 as dependency for these.

In every place Pandas gets imported, both versions would need to be handled (a major version usually indicates breaking changes, so some exception handling is required there to handle both code paths).

**Reply by loldenburg** [reply]
@Ryuno-Ki I did read the whole chapter, but I did not find anything that would match my requirement. The system and os environment is the same for all python files, but for some I want to load pandas 0.25, for some pandas 1.00. But I probably simply don't get it.

**Reply by loldenburg** [reply]
@Ryuno-Ki I now checked this with my supertech nerd and confirm that the flags are not going to cut it. 

"No, these are markers that can tell e.g. that this dependency is only required for Windows OS, or that the package can be only used with Python <= 2.7 or smth similar.

It'll not fix the runtime library dependency resolution strategy."

**Reply by Ryuno-Ki** [reply]
Thanks. What a bummer!

**Comment by loldenburg** [comment]
It’s definitely on my high prio to do list to support pandas 1+  Thanks
again!
